Objects/typeobject.c

```diff
@@ -3741,6 +3741,76 @@ type_is_gc(PyTypeObject *type)
     return type->tp_flags & Py_TPFLAGS_HEAPTYPE;
 }
 
+
+// TODO: MM: Move the following three functions?
+static int
+is_typing_name(PyObject *obj, char *name)
+{
+    PyTypeObject *type = Py_TYPE(obj);
+    if (strcmp(type->tp_name, name) != 0) {
+        return 0;
+    }
+    PyObject *module = PyObject_GetAttrString((PyObject *)type, "__module__");
+    if (module == NULL) {
+        return -1;
+    }
+    int res = PyUnicode_Check(module)
+        && _PyUnicode_EqualToASCIIString(module, "typing");
+    Py_DECREF(module);
+    return res;
+}
+
+static int
+is_typevar(PyObject *obj)
+{
+    return is_typing_name(obj, "TypeVar");
+}
+
+static int
+is_genericalias(PyObject *obj)
+{
+    return is_typing_name(obj, "GenericAlias");
+}
+
+
+static PyObject *
+type_or(PyTypeObject* self, PyObject* param) {
+    // Check param is a PyType or GenericAlias
+    if ((param == NULL) ||
+        (
+         (param != Py_None) &
+         ! is_genericalias(param) &&
+         ! is_typevar(param) &&
+         (PyObject_IsInstance(param, (PyObject *) &PyType_Type) != 1) &&
+          (PyObject_IsInstance(param, (PyObject *) &Py_UnionType) != 1)
+        )
+            ) {
+        PyErr_SetString(PyExc_TypeError, "'type' expected");
+        return NULL;
+    }
+
+    PyObject *tuple;
+    if (PyObject_IsInstance((PyObject *)self, (PyObject *) &Py_UnionType)) {
+        tuple = Py_Union_AddToTuple((PyObject *)self, param, 0);
+    } else if (PyObject_IsInstance(param, (PyObject *) &Py_UnionType)) {
+        tuple = Py_Union_AddToTuple(param, (PyObject *)self, 1);
+    } else {
+        PyObject *param_type = param == Py_None ? (PyObject *)Py_TYPE(param) : param;
+        PyTypeObject *self_type = (PyObject *)self == Py_None ? Py_TYPE(self) : self;
+        tuple=PyTuple_Pack(2, self_type, param_type);
+        Py_DECREF(param_type);
+        Py_DECREF(self_type);
+    }
+
+    PyObject *newUnionType=Py_Union(tuple);
+    Py_DECREF(tuple);
+    return newUnionType;
```

Comment:
There was a test case in the original Implementation: ()

That looked like:

`self.assertEqual(None | int, typing.Union[int, None])`

(This test currently fails with this implementation). Building off the above example, should it also be true that 
`self.assertEqual(int | bool | str, typing.Union[int, str, bool])`

If so, I think a potential solution would be to sort the arguments when we create the `Py_Union` type. Not sure of the best way to implement that sorting functionality.